i18n/src/main/scala/io/udash/i18n/TranslationKey.scala

```diff
@@ -1,141 +1,134 @@
 }
 
 object TranslationKey {
   def key(key: String): TranslationKey0 =
-    new TranslationKey0(key)
+    SimpleTranslationKey0(key)
 
   def key1[T](key: String): TranslationKey1[T] =
-    new TranslationKey1[T](key)
+    TranslationKey1[T](key)
 
   def key2[T1, T2](key: String): TranslationKey2[T1, T2] =
-    new TranslationKey2[T1, T2](key)
+    TranslationKey2[T1, T2](key)
 
   def key3[T1, T2, T3](key: String): TranslationKey3[T1, T2, T3] =
-    new TranslationKey3[T1, T2, T3](key)
+    TranslationKey3[T1, T2, T3](key)
 
   def key4[T1, T2, T3, T4](key: String): TranslationKey4[T1, T2, T3, T4] =
-    new TranslationKey4[T1, T2, T3, T4](key)
+    TranslationKey4[T1, T2, T3, T4](key)
 
   def key5[T1, T2, T3, T4, T5](key: String): TranslationKey5[T1, T2, T3, T4, T5] =
-    new TranslationKey5[T1, T2, T3, T4, T5](key)
+    TranslationKey5[T1, T2, T3, T4, T5](key)
 
   def key6[T1, T2, T3, T4, T5, T6](key: String): TranslationKey6[T1, T2, T3, T4, T5, T6] =
-    new TranslationKey6[T1, T2, T3, T4, T5, T6](key)
+    TranslationKey6[T1, T2, T3, T4, T5, T6](key)
 
   def key7[T1, T2, T3, T4, T5, T6, T7](key: String): TranslationKey7[T1, T2, T3, T4, T5, T6, T7] =
-    new TranslationKey7[T1, T2, T3, T4, T5, T6, T7](key)
+    TranslationKey7[T1, T2, T3, T4, T5, T6, T7](key)
 
   def key8[T1, T2, T3, T4, T5, T6, T7, T8](key: String): TranslationKey8[T1, T2, T3, T4, T5, T6, T7, T8] =
-    new TranslationKey8[T1, T2, T3, T4, T5, T6, T7, T8](key)
+    TranslationKey8[T1, T2, T3, T4, T5, T6, T7, T8](key)
 
   def key9[T1, T2, T3, T4, T5, T6, T7, T8, T9](key: String): TranslationKey9[T1, T2, T3, T4, T5, T6, T7, T8, T9] =
-    new TranslationKey9[T1, T2, T3, T4, T5, T6, T7, T8, T9](key)
+    TranslationKey9[T1, T2, T3, T4, T5, T6, T7, T8, T9](key)
 
   def keyX(key: String): TranslationKeyX =
-    new TranslationKeyX(key)
+    TranslationKeyX(key)
 
   def untranslatable(key: String): TranslationKey0 =
-    new Untranslatable(key)
+    Untranslatable(key)
 
-  private[i18n] final class ReducedTranslationKey(key: String, val argv: Any*) extends TranslationKey0(key) {
-    override def apply()(implicit provider: TranslationProvider, lang: Lang): Future[Translated] = {
+  private[i18n] final case class SimpleTranslationKey0(key: String) extends TranslationKey0
+
+  private[i18n] final case class ReducedTranslationKey(key: String, argv: Any*) extends TranslationKey0 {
+    override def apply()(implicit provider: TranslationProvider, lang: Lang): Future[Translated] =
       provider.translate(key, argv: _*)
-    }
+    // default toString puts argv inside "WrappedArray()"
+    override def toString(): String = s"$productPrefix($key,${argv.mkString(",")})"
   }
 
-  private[i18n] final class Untranslatable(key: String) extends TranslationKey0(key) {
+  private [i18n] object ReducedTranslationKey {
+    // note: serialization loses information, as argv is converted to strings
```

Comment:
Since we control all instantiation of `ReducedTranslationKey`s, maybe we could just change `argv` type to `String*` and render this comment useless? I don't think we're particularly afraid of making `.toString` calls too strict.

Reply:
I did not do it earlier, as `TranslationProvider.translate` still receives `Any*`, so custom implementations could possibly do some fancy number formatting and the like, which string conversion would foil. Still, I agree that always doing the conversion is preferable to losing info on serialization.